src/httpserver.cpp

```diff
@@ -417,361 +286,708 @@ static void HTTPWorkQueueRun(WorkQueue<HTTPClosure>* queue, int worker_num)
     queue->Run();
 }
 
-/** libevent event log callback */
-static void libevent_log_cb(int severity, const char *msg)
+void RegisterHTTPHandler(const std::string &prefix, bool exactMatch, const HTTPRequestHandler &handler)
 {
-    BCLog::Level level;
-    switch (severity) {
-    case EVENT_LOG_DEBUG:
-        level = BCLog::Level::Debug;
-        break;
-    case EVENT_LOG_MSG:
-        level = BCLog::Level::Info;
-        break;
-    case EVENT_LOG_WARN:
-        level = BCLog::Level::Warning;
-        break;
-    default: // EVENT_LOG_ERR and others are mapped to error
-        level = BCLog::Level::Error;
-        break;
+    LogDebug(BCLog::HTTP, "Registering HTTP handler for %s (exactmatch %d)\n", prefix, exactMatch);
+    LOCK(g_httppathhandlers_mutex);
+    pathHandlers.emplace_back(prefix, exactMatch, handler);
+}
+
+void UnregisterHTTPHandler(const std::string &prefix, bool exactMatch)
+{
+    LOCK(g_httppathhandlers_mutex);
+    std::vector<HTTPPathHandler>::iterator i = pathHandlers.begin();
+    std::vector<HTTPPathHandler>::iterator iend = pathHandlers.end();
+    for (; i != iend; ++i)
+        if (i->prefix == prefix && i->exactMatch == exactMatch)
+            break;
+    if (i != iend)
+    {
+        LogDebug(BCLog::HTTP, "Unregistering HTTP handler for %s (exactmatch %d)\n", prefix, exactMatch);
+        pathHandlers.erase(i);
     }
-    LogPrintLevel(BCLog::LIBEVENT, level, "%s\n", msg);
 }
 
-bool InitHTTPServer(const util::SignalInterrupt& interrupt)
+namespace http_bitcoin {
+using util::Split;
+
+std::optional<std::string_view> HTTPHeaders::Find(const std::string key) const
 {
-    if (!InitHTTPAllowList())
-        return false;
+    const auto it = m_map.find(key);
+    if (it == m_map.end()) return std::nullopt;
+    return std::string_view(it->second);
```

Comment:
This returns a reference (or a "pointer") to the outside world of an element from the private `m_map` member. That map is not immutable. What happens if the strings it contains get moved around if the map is internally resized when adding new entries? Also there is the `HTTPHeaders::Remove()` method which deletes entries from the map, rendering any returned pointers by `Find()` as dangling.

54a36093f7 `http: using string_view to avoid unnecessary copy in Headers` -- if there is no any measurable performance gain from this commit then it would be safer to drop it. If there is a measurable performance gain then hmm...